src/backups.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of backups."""
+
+import logging
+from typing import Tuple
+
+from constants import SERVER_CONFIG_USERNAME, SERVER_CONFIG_PASSWORD_KEY
+from charms.mysql.v0.mysql import (
+    MySQLGetMemberStateError,
+    MySQLSetInstanceOfflineModeError,
+    MySQLSetInstanceOptionError,
+)
+from mysql_k8s_helpers import MySQLExecuteBackupScriptError
+from ops.charm import ActionEvent, CharmBase
+from ops.framework import Object
+
+logger = logging.getLogger(__name__)
+
+MYSQL_BACKUPS = "mysql-backups"
+
+
+class MySQLBackups(Object):
+    """Encapsulation of backups for MySQL."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, MYSQL_BACKUPS)
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.perform_backup_action, self._on_perform_backup)
+
+    def _on_perform_backup(self, event: ActionEvent) -> None:
+        """Perform backup action."""
+        logger.info("A backup has been requested on unit")
+
+        s3_bucket = event.params.get("s3-bucket")
+        s3_path = event.params.get("s3-path")
+        s3_access_key = event.params.get("s3-access-key")
+        s3_secret_key = event.params.get("s3-secret-key")
+
+        if not s3_bucket or not s3_path or not s3_access_key or not s3_secret_key:
```

Comment:
Can we create a function which will check if s3 credentials are valid?
It is simple with the following command
```
xbstream -C /tmp -c /some/not-empty-file \
    | xbcloud put \
           --curl-retriable-errors=7 \
           --insecure \
           --storage=s3 \
           --parallel=10 \
           --md5 \
           --s3-bucket="$S3_BUCKET" \
           "$S3_BUCKET_PATH.not-empty-file"
```

Reply:
discussed offline: xbcloud exits with error code 1 when uploading anything but a backup file. instead, we decided to download boto3 and use it to upload to S3